src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -0,0 +1,526 @@
+            ),
+            dependencies=MavenCoordinates(MavenCoord(coord=d) for d in entry["dependencies"]),
+            file_digest=FileDigest(
+                fingerprint=entry["file_digest"]["fingerprint"],
+                serialized_bytes_length=entry["file_digest"]["serialized_bytes_length"],
+            ),
+        )
+
+    def to_json_dict(self):
+        """Export this CoursierLockfileEntry to a JSON object."""
+
+        return dict(
+            coord=self.coord.coord,
+            directDependencies=[coord.coord for coord in self.direct_dependencies],
+            dependencies=[coord.coord for coord in self.dependencies],
+            file_name=self.file_name,
+            file_digest=dict(
+                fingerprint=self.file_digest.fingerprint,
+                serialized_bytes_length=self.file_digest.serialized_bytes_length,
+            ),
+        )
+
+
+@dataclass(frozen=True)
+class CoursierResolvedLockfile:
+    """An in-memory representation of Pants' Coursier lockfile format."""
+
+    entries: Tuple[CoursierLockfileEntry, ...]
+
+    @classmethod
+    def from_json_dict(cls, lockfile) -> CoursierResolvedLockfile:
+        """Construct a CoursierResolvedLockfile from its JSON dictionary representation."""
+
+        return CoursierResolvedLockfile(
+            entries=tuple(CoursierLockfileEntry.from_json_dict(dep) for dep in lockfile)
+        )
+
+    def to_json(self) -> bytes:
+        """Export this CoursierResolvedLockfile to human-readable JSON.
+
+        This JSON is intended to be checked in to the user's repo as a hermetic snapshot of a
+        Coursier resolved JVM classpath.
+        """
+        return json.dumps([entry.to_json_dict() for entry in self.entries], indent=4).encode(
+            "utf-8"
+        )
+
+
+@rule(level=LogLevel.DEBUG)
+async def coursier_resolve_lockfile(
+    coursier: Coursier,
+    maven_requirements: MavenRequirements,
+) -> CoursierResolvedLockfile:
+    """Run `coursier fetch ...` against a list of Maven coordinates and capture the result.
+
+    This rule does two things in a single Process invocation:
+
+        * Runs `coursier fetch` to let Coursier do the heavy lifting of resolving
+          dependencies and downloading resolved artifacts (jars, etc).
+        * Copies the resolved artifacts into the Process output directory, capturing
+          the artifacts as content-addressed `Digest`s.
+
+    It's important that this happens in the same process, since the process isn't
+    guaranteed to run on the same machine as the rule, nor is a subsequent process
+    invocation.  This guarantees that whatever Coursier resolved, it was fully
+    captured into Pants' content addressed artifact storage.
+
+    Note however that we still get the benefit of Coursier's "global" cache if it
+    had already been run on the machine where the `coursier fetch` runs, so rerunning
+    `coursier fetch` tends to be fast in practice.
```

Comment:
Why is this? Ik you asked about append_only_caches, but doesn't look like it's being used.

Reply:
Coursier by default uses its own global cache in the user's home directory.  So if it's already been run on the machine where the process executes, it's likely to get cache hits there and avoid expensive network fetches